**MAESTRÍA EN INTELIGENCIA ARTIFICIAL APLICADA**

**Curso: TC5053 - Ciencia y analítica de datos**

Tecnológico de Monterrey

Prof Grettel Barceló Alonso

**Semana 3**
Bases, almacenes y manipulación de datos

---

*   NOMBRE: Kléber Mauricio Chiriboga Iñiguez
*   MATRÍCULA: A01841624


---

En esta actividad usarás una base de datos relacional basada en el informe de participación y la lista del top 10 de Netflix. Incluye películas y programas de televisión, así como información sobre temporadas, métricas de visualización, fechas de estreno, duración y más, organizada en las siguientes tablas:

* `movie`: Información general de las películas.

* `tv_show`: Información general de los programas de televisión.

* `season`: Datos de las temporadas asociadas a cada programa de TV.

* `view_summary`: Métricas de visualización y rendimiento de películas o temporadas.

Revisa con detalle su esquema para que comprendas cómo se relacionan las tablas anteriores.

**NOTA IMPORTANTE:** Asegúrate de responder *explícitamente* todos los cuestionamientos.


`PyMySQL` es una librería escrita en Python puro que funciona como conector (*driver*) para motores de bases de datos MySQL, permitiendo abrir conexiones, ejecutar consultas SQL y recuperar resultados directamente desde programas en Python.

In [1]:
pip install pymysql

`SQLAlchemy` es una librería de Python que facilita la interacción con bases de datos y permite mantener un pool de conexiones eficiente, gestionar *commits* y *rollbacks* de forma automática y asegurar que múltiples conexiones simultáneas se manejen de manera segura, incluso cuando se ejecutan consultas SQL “puras”

In [2]:
# Importa las librerías necesarias
import pymysql
import sqlalchemy as sqla
import pandas as pd

Se crea una conexión (`conn`) para luego invocar declaraciones SQL.

In [15]:
# motor+driver://usuarioBD:clave@ipHostDBMS:puerto/esquema
# pool_recycle controla el tiempo máximo de vida de una conexión en el pool (3600 segundos = 1 hora)
db = sqla.create_engine('mysql+pymysql://mnaTC5053User:mnaTC5053Pass!@20.127.169.219:3306/Netflix', pool_recycle=3600)
conn = db.connect()

Para que tus consultas sean más legibles y fáciles de mantener, puedes usar este formato multilínea con comillas triples y `sqla.text()`. Por ejemplo:

```
query = sqla.text("""
  SELECT ---
  FROM ---
  WHERE ---
""")
pd.read_sql(query, conn)
```

1.	Extrae toda la información de las películas que duran más de 5 horas.

In [17]:
query = sqla.text("""
SELECT *
FROM movie
WHERE runtime > 300
""")
pd.read_sql(query, conn)

,id,created_date,modified_date,available_globally,locale,original_title,release_date,runtime,title
0,5793,2024-01-01,2024-01-01,b'\x00',None,日本統一シリーズ: 映画シリーズ,None,3892,Nihontouitsu Series: Film Series
1,5794,2024-01-01,2024-01-01,b'\x00',None,釣りバカ日誌: 映画シリーズ,None,2120,Free and Easy Series: Film Series
2,5874,2024-01-01,2024-01-01,b'\x01',None,None,2021-08-06,312,Navarasa: Limited Series
3,9729,2024-01-01,2024-01-01,b'\x00',None,織田同志会 織田征仁: 映画シリーズ,None,710,Seiji Oda: Film Series
4,9730,2024-01-01,2024-01-01,b'\x00',None,キングダム～首領になった男～: 映画シリーズ,None,427,Kingdom ~ The Man Who Became the Top ~: Film S...


2.	¿Cuál es el porcentaje de películas disponibles únicamente en EU en relación con el total, excluyendo los valores `NULL`? La consulta SQL debe entregar directamente el porcentaje final. No se deben devolver resultados parciales para realizar el cálculo en Python.

In [18]:
query = sqla.text("""
SELECT 100 * SUM(available_globally = 0) / COUNT(available_globally) AS porcentaje_solo_eu
FROM movie
""")
pd.read_sql(query, conn)

,porcentaje_solo_eu
0,83.5717


3.	¿Cuáles son los idiomas o regiones originales en la tabla de películas?
* ¿Cuántos registros tienen el campo `locale` con valor `NULL`? (NULL en SQL ⇔ None en Python)

In [19]:
query = sqla.text("""
SELECT locale, COUNT(*) AS total
FROM movie
GROUP BY locale
""")
pd.read_sql(query, conn)

,locale,total
0,None,11255
1,en,532


4.	Asumiendo que los valores `NULL` en `locale` corresponden a otro idioma (diferente del inglés), el título original de la película NO debería coincidir con el título principal en dichos registros.
* Determina cuántas películas tienen títulos diferentes en estos dos campos (`title` y `original_title`).
*  ¿Coinciden los resultados (cantidad de `NULL` y títulos diferentes)? Si no es así, identifica qué características tienen los registros restantes.
* Finalmente, concluye si la suposición de que los valores `NULL` en `locale` indican que la película está en otro idioma es válida.

In [20]:
query = sqla.text("""
SELECT COUNT(*) AS titulos_diferentes
FROM movie
WHERE title <> original_title
""")
pd.read_sql(query, conn)

,titulos_diferentes
0,3947


In [21]:
query = sqla.text("""
SELECT COUNT(*) AS registros_restantes,
       SUM(original_title IS NULL) AS sin_titulo_original,
       SUM(title = original_title) AS titulos_iguales
FROM movie
WHERE locale IS NULL
  AND (original_title IS NULL OR title = original_title)
""")
pd.read_sql(query, conn)

,registros_restantes,sin_titulo_original,titulos_iguales
0,7308,7291.0,17.0


Los resultados no coinciden. Hay 11,255 registros con locale NULL, pero solo 3,947 tienen un título diferente al original.

De los 7,308 que quedan, 7,291 no tienen título original (está en NULL) y 17 tienen el mismo título en los 2 campos.

Por lo cual la suposición no es válida. Que locale sea NULL no significa que la película esté en otro idioma. En la mayoría de los casos solo falta el dato.

5.	Determina el título de la película que ha permanecido más tiempo en el top 10.

In [22]:
query = sqla.text("""
SELECT m.title, v.cumulative_weeks_in_top10
FROM movie m
JOIN view_summary v ON v.movie_id = m.id
ORDER BY v.cumulative_weeks_in_top10 DESC
LIMIT 1
""")
pd.read_sql(query, conn)

,title,cumulative_weeks_in_top10
0,The Boss Baby,22


6.	Identifica los 5 programas de TV con mayor cantidad de temporadas.

In [23]:
query = sqla.text("""
SELECT t.title, COUNT(*) AS temporadas
FROM tv_show t
JOIN season s ON s.tv_show_id = t.id
GROUP BY t.id, t.title
ORDER BY temporadas DESC, t.title
LIMIT 5
""")
pd.read_sql(query, conn)

,title,temporadas
0,Grey's Anatomy,20
1,Naruto Shippuden,20
2,Heartland (2007),17
3,It's Always Sunny in Philadelphia,16
4,NCIS,15


Nota: hay un empate en el quinto lugar. NCIS, Raw y Supernatural (2005) tienen 15 temporadas cada uno. La consulta solo uno porque se pidió un máximo de 5 filas.

7.	¿Cuáles son los intervalos de fechas de los resúmenes en la tabla `view_summary` de los períodos (`duration`) semestrales?

In [24]:
query = sqla.text("""
SELECT DISTINCT start_date, end_date
FROM view_summary
WHERE duration = 'SEMI_ANNUALLY'
ORDER BY start_date
""")
pd.read_sql(query, conn)

,start_date,end_date
0,2023-07-01,2023-12-31
1,2024-01-01,2024-06-30


8.	Ordena las temporadas de *Grey's Anatomy* según la cantidad de vistas registradas en el primer período semestral de 2024.
* ¿Cómo interpretarías los resultados obtenidos?

In [25]:
query = sqla.text("""
SELECT s.season_number, s.title, v.views
FROM tv_show t
JOIN season s ON s.tv_show_id = t.id
JOIN view_summary v ON v.season_id = s.id
WHERE t.title = "Grey's Anatomy"
  AND v.duration = 'SEMI_ANNUALLY'
  AND v.start_date = '2024-01-01'
ORDER BY v.views DESC, s.season_number
""")
pd.read_sql(query, conn)

,season_number,title,views
0,1,Grey's Anatomy: Season 1,3600000
1,2,Grey's Anatomy: Season 2,3100000
2,3,Grey's Anatomy: Season 3,2900000
3,4,Grey's Anatomy: Season 4,2900000
4,5,Grey's Anatomy: Season 5,2900000
5,6,Grey's Anatomy: Season 6,2800000
6,7,Grey's Anatomy: Season 7,2700000
7,8,Grey's Anatomy: Season 8,2600000
8,9,Grey's Anatomy: Season 9,2500000
9,10,Grey's Anatomy: Season 10,2400000


La temporada 1 es la más vista, con 3.6 millones. Después las vistas van bajando poco a poco hasta 1.7 millones en las temporadas 17 y 18.

Esto indica que mucha gente empieza la serie desde el inicio, pero no todos llegan a las últimas temporadas.

Hay dos casos distintos. La temporada 19 tiene más vistas que las temporadas 15 a 18. La temporada 20 tiene muy pocas (100 mil), posiblemente porque estuvo disponible poco tiempo en ese semestre.

9.	Todas las consultas anteriores podrían hacerse también con la lógica relacional implementada en Pandas, que permite replicar la mayoría de las operaciones de SQL. Si los dataframes se han llenado como sigue, resuelve la consulta 8 utilizando las funciones de Pandas.

In [26]:
tv_show = pd.read_sql(sqla.text('SELECT * FROM tv_show'), conn)
season = pd.read_sql(sqla.text('SELECT * FROM season'), conn)
view_summary = pd.read_sql(sqla.text('SELECT * FROM view_summary'), conn)

In [27]:
grey = tv_show[tv_show['title'] == "Grey's Anatomy"]
temporadas = grey.merge(season, left_on='id', right_on='tv_show_id', suffixes=('_show', '_season'))
resumen = temporadas.merge(view_summary, left_on='id_season', right_on='season_id')

filtro = (resumen['duration'] == 'SEMI_ANNUALLY') & (resumen['start_date'].astype(str) == '2024-01-01')
resultado = resumen[filtro][['season_number', 'title_season', 'views']]
resultado.sort_values(['views', 'season_number'], ascending=[False, True])

,season_number,title_season,views
0,1.0,Grey's Anatomy: Season 1,3600000
2,2.0,Grey's Anatomy: Season 2,3100000
4,3.0,Grey's Anatomy: Season 3,2900000
8,4.0,Grey's Anatomy: Season 4,2900000
6,5.0,Grey's Anatomy: Season 5,2900000
10,6.0,Grey's Anatomy: Season 6,2800000
12,7.0,Grey's Anatomy: Season 7,2700000
14,8.0,Grey's Anatomy: Season 8,2600000
16,9.0,Grey's Anatomy: Season 9,2500000
18,10.0,Grey's Anatomy: Season 10,2400000


`MySQL` es un sistema de gestión de bases de datos relacional, pero desde Python también es posible extraer información de bases de datos no relacionales, como `Firestore`, `MongoDB` o `Cassandra`, utilizando conectores o integraciones específicas. Esto permite combinar datos de diferentes fuentes según las necesidades del análisis o la aplicación.

En el siguiente ejercicio usarás un ejemplo con `Firestore` desde Python. Para ello utilizarás los módulos `credentials` y `firestore` de la biblioteca `firebase_admin`.

In [28]:
import firebase_admin
from firebase_admin import credentials
from firebase_admin import firestore

En `Firestore`, a diferencia de `MySQL` donde se utiliza un usuario y contraseña para conectarse, la autenticación se realiza mediante un archivo JSON que almacena las credenciales necesarias para acceder a la base de datos. Este archivo contiene las claves y la información de configuración que permiten a Python establecer la conexión de manera segura.

In [29]:
# Debes descargar el archivo de credenciales "consultancy.json" (disponible en las instrucciones de Canvas) y ubicarte en la carpeta donde lo almacenaste
import os
DIR = "/content"
os.chdir(DIR)

In [30]:
# consultancy.json almacena la clave privada para autenticar una cuenta y autorizar el acceso a los servicios
# A través de la función Certificate(), se regresa una credencial inicializada, que puedes utilizar para crear una nueva instancia de la aplicación
cred = credentials.Certificate('consultancy.json')
firebase_admin.initialize_app(cred)
db = firestore.client()

10.	Investiga cómo leer la colección `EMPLOYEE` y mostrar su contenido en un dataframe. Asegúrate de incluir el `id` en el resultado

In [31]:
docs = db.collection('EMPLOYEE').stream()

datos = []
for doc in docs:
    fila = {'id': doc.id}
    fila.update(doc.to_dict())
    datos.append(fila)

pd.DataFrame(datos)

,id,emp_fname,emp_lname,emp_hiredate
0,8LcLuxVHGAd2d9IQc5jF,David,Senior,1989-07-12 06:00:00+00:00
1,Fzd60D6Z2CU4n0wVV8YN,William,Smithfield,2004-06-04 05:00:00+00:00
2,lX5xuQ5w3i6ib2ExccWY,John,News,2000-11-08 06:00:00+00:00
3,yocFj2lichOkbAj9NBfp,June,Arbough,1996-12-01 06:00:00+00:00


In [32]:
firebase_admin.delete_app(firebase_admin.get_app())

---

**Declaración de uso de IA**

Si aplica, deberá indicarse la herramienta y el modelo empleado en la entrega, así como la finalidad de su uso (generación de código / depuración / optimización).

Por ejemplo:

*   OpenAI. (2026). *ChatGPT (basado en GPT-4)* [Modelo de lenguaje grande], utilizado para generación de código y depuración. https://chat.openai.com/

**Declaración:**

*   Anthropic. (2026). *Claude (modelo claude-fable-5-1)* [Modelo de lenguaje grande], utilizado para apoyo generación y revisión de código y análisis. https://claude.ai/

---